### PR TITLE
Fix hellfire save games corrupting the game state

### PR DESCRIPTION
**Pull request by AJenbo**
Fixes https://github.com/diasurgical/devilutionX/issues/1367

### CODE REVIEW

File: Source/pfile.cpp [hunk 1]
```diff
@@ -153,7 +153,8 @@ static HANDLE pfile_open_save_archive(DWORD save_num)
 
 static void pfile_SFileCloseArchive(HANDLE hsArchive)
 {
-	SFileCloseArchive(hsArchive);
+	if (hsArchive)
```
Comment (glebm) [inline Source/pfile.cpp:156]: Why do we call this function if the archive isn't open? I think we should `assert(hsArchive != NULL)` here instead

File: Source/pfile.cpp [hunk 2]
```diff
@@ -225,6 +226,8 @@ bool pfile_ui_set_hero_infos(bool (*ui_add_hero_info)(_uiheroinfo *))
 
 				UnPackPlayer(&pkplr, 0, false);
 
+				pfile_SFileCloseArchive(archive);
```
Comment (glebm) [inline Source/pfile.cpp:229]: Let's change the argument to `pfile_SFileCloseArchive` to `HANDLE *` and make `pfile_SFileCloseArchive` set it to NULL instead

File: Source/pfile.cpp [hunk 3]
```diff
@@ -225,6 +226,8 @@ bool pfile_ui_set_hero_infos(bool (*ui_add_hero_info)(_uiheroinfo *))
 
 				UnPackPlayer(&pkplr, 0, false);
 
+				pfile_SFileCloseArchive(archive);
+				archive = NULL;
 				LoadHeroItems(&plr[0]);
 				RemoveEmptyInventory(0);
 				CalcPlrInv(0, false);
```
Comment (glebm) [inline Source/pfile.cpp:233]: Why is there another call `pfile_SFileCloseArchive` on line 238?
If it is needed, it should be removed from there and moved to a more appropriate place

Comment (AJenbo) [reply Source/pfile.cpp:233]: If unpacking was a success or not.

Comment (qndel) [reply Source/pfile.cpp:233]: Because LoadHeroItems opens the archive again and failed earier because it required a closed archive to open it 🤪

Comment (glebm) [reply Source/pfile.cpp:233]: Instead of calling `close` unconditionally, it should only be called if `open` was called

Comment (AJenbo) [reply Source/pfile.cpp:233]: I think i will save that for a later refactor. To much global stuff going on for me to feel safe about shifting stuff around for a hotfix.

File: Source/pfile.cpp (hunk 1 above)
Comment (AJenbo) [reply Source/pfile.cpp:156]: We need to be able to safely call it multiple times, or more carefully track the state.
Or I don't know how to do this atm, feel free to send a PR that does this nicer.

**Review by glebm — APPROVED** [review]

**Pull request merged**